### PR TITLE
kubelet/rkt: garbage collect systemd service files in GarbageCollect().

### PR DESCRIPTION
**Pull request by yifan-gu**
Not removing the service file in KillPod() because when starting the pod, we need a way to tell if the restart count should be increased. And currently the restart count is stored in the systemd service file.

I'm open to other solution. Maybe we should create a file to store the info under `/var/lib/kubelet/pod/$UID/container/`, and this should be helpful for docker runtime as well. It should simplify the computation of the restart count, such as [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockertools/manager.go#L461)

### CODE REVIEW

**Comment by yifan-gu** [comment]
/cc @dchen1107 @yujuhong @timothysc @vishh 
For some reason seems I can't @kubernetes/goog-node ? :(

**Comment by yifan-gu** [comment]
Pending on https://github.com/kubernetes/kubernetes/pull/15051

**Comment by yujuhong** [comment]
I think you wanted another Tim St Clair. cc'ing the right one @timstclair  :)



> For some reason seems I can't @kubernetes/goog-node ? :(

I noticed this too. Not sure what's criteria. Will check later.

**Comment by yujuhong** [comment]
> I'm open to other solution. Maybe we should create a file to store the info under /var/lib/kubelet/pod/$UID/container/, and this should be helpful for docker runtime as well. It should simplify the computation of the restart count, such as here

We had the discussion recently. The restart count code is problematic and I have fixed it (*too*) many times :(
Basically we can either add kubelet checkpointing to persist various specs/states or push down to container runtimes to persistent the restart count (#15089: write restart count and other things to docker labels)
I had the same concern of relying on the existence of the previous container to retrieve the restart count, which may complicate the logic. However, checkpointing is a bigger project to take on, and people have concerns about it too. I think our short-term solution is docker labels (i.e., rkt service files), but perhaps this would change? @kubernetes/goog-node

File: pkg/kubelet/rkt/rkt.go [hunk 1]
```diff
@@ -806,13 +806,20 @@ func (r *Runtime) KillPod(pod *api.Pod, runningPod kubecontainer.Pod) error {
 		r.containerRefManager.ClearRef(id)
 	}
 
+	// Touch the systemd service file to update the mod time so it will
+	// not be garbage collected too soon.
+	if err := os.Chtimes(serviceFilePath(serviceName), time.Now(), time.Now()); err != nil {
+		glog.Errorf("rkt: Failed to change the modification time of the service file %v: %v", serviceName, err)
```
Comment (timstclair) [inline pkg/kubelet/rkt/rkt.go]: (nit) `%q` for string strings (I just learned this)

File: pkg/kubelet/rkt/rkt.go [hunk 2]
```diff
@@ -1092,6 +1099,38 @@ func (r *Runtime) GarbageCollect() error {
...
 		return err
 	}
+
+	// GC all inactive systemd service files.
+	units, err := r.systemd.ListUnits()
+	if err != nil {
+		glog.Errorf("rkt: Failed to list units: %v", err)
+		return err
+	}
+	runningUnits := make(map[string]struct{})
```
Comment (timstclair) [inline pkg/kubelet/rkt/rkt.go]: (optional) Consider using a [`StringSet`](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/sets/set.go#L29)

File: pkg/kubelet/rkt/rkt.go [hunk 3]
```diff
@@ -1092,6 +1099,38 @@ func (r *Runtime) GarbageCollect() error {
 		glog.Errorf("rkt: Failed to gc: %v", err)
 		return err
```
Comment (timstclair) [inline pkg/kubelet/rkt/rkt.go]: If this fails, should you still try to garbage collect the service files?

Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: I think so. Thanks :)

File: pkg/kubelet/rkt/rkt.go (hunk 2 above)
Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: ack

File: pkg/kubelet/rkt/rkt.go (hunk 1 above)
Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: yeah. lazyness...

File: pkg/kubelet/rkt/rkt.go [hunk 4]
```diff
@@ -1086,12 +1094,43 @@ func (r *Runtime) GetContainerLogs(pod *api.Pod, containerID string, logOptions
...
+		}
+	}
+
+	files, err := ioutil.ReadDir(systemdServiceDir)
+	if err != nil {
+		glog.Errorf("rkt: Failed to read the systemd service directory: %v", err)
 		return err
 	}
+	for _, f := range files {
+		if strings.HasPrefix(f.Name(), kubernetesUnitPrefix) {
```
Comment (yujuhong) [inline pkg/kubelet/rkt/rkt.go]: nit: reduce the nested `if` levels by using `continue` here? This is more of a personal preference, so it's up to you.

**Comment by yujuhong** [comment]
@timstclair, since you have already started reviewing this PR, I assigned it to you :)

**Comment by yifan-gu** [comment]
Rebased

**Comment by yujuhong** [comment]
> Rebased

Not enough...

**Comment by yifan-gu** [comment]
ping?

File: pkg/kubelet/rkt/rkt.go [hunk 5]
```diff
@@ -1079,12 +1087,43 @@ func (r *Runtime) GetContainerLogs(pod *api.Pod, containerID kubecontainer.Conta
...
+	}
+
+	files, err := ioutil.ReadDir(systemdServiceDir)
+	if err != nil {
+		glog.Errorf("rkt: Failed to read the systemd service directory: %v", err)
+		return err
+	}
+	for _, f := range files {
+		if strings.HasPrefix(f.Name(), kubernetesUnitPrefix) {
+			if _, ok := runningUnits[f.Name()]; !ok {
```
Comment (timstclair) [inline pkg/kubelet/rkt/rkt.go]: (nit) `if runningUnits.Has(f.Name()) { ...`

File: pkg/kubelet/rkt/rkt.go [hunk 6]
```diff
@@ -1079,12 +1087,43 @@ func (r *Runtime) GetContainerLogs(pod *api.Pod, containerID kubecontainer.Conta
...
+		}
+	}
+
+	files, err := ioutil.ReadDir(systemdServiceDir)
+	if err != nil {
+		glog.Errorf("rkt: Failed to read the systemd service directory: %v", err)
+		return err
+	}
+	for _, f := range files {
+		if strings.HasPrefix(f.Name(), kubernetesUnitPrefix) {
```
Comment (timstclair) [inline pkg/kubelet/rkt/rkt.go]: (nit) I'd prefer to combine these three ifs to reduce indentation levels. If that is too unwieldy, you could also consider inverting them with continues.

**Comment by timstclair** [comment]
LGTM. Sorry I lost track of this.

File: pkg/kubelet/rkt/rkt.go (hunk 6 above)
Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: thanks, just realized the first if is redundant :)

**Comment by yifan-gu** [comment]
> LGTM. Sorry I lost track of this.

No worries, thanks for the review, comments addressed :)

**Pull request merged**